### PR TITLE
Enable proton revision for checkpoint

### PR DESCRIPTION
**Pull request by yl-lisen**
This is phase 1 for #421 

Please write user-readable short description of the changes:
- Enable proton revision for checkpoint
- revise partial serialize/deserialize interface to support versioning

### CODE REVIEW

File: src/Common/serde.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+template <typename T>
+concept HasGetVersionFromRevision = requires(T && a) {
+    {
+        a.getVersionFromRevision(ProtonRevision::getVersionRevision())
+    } -> std::convertible_to<VersionType>;
+};
+
+template <typename WB, VersionedSerializable<WB> S>
+void ALWAYS_INLINE serialize(const S & s, WB && wb, std::optional<VersionType> version = std::nullopt)
```
Comment (chenziliang) [inline src/Common/serde.h]: This appears too complex and has too many branches. I like to recommend
1) make code simple to understand 
2) make code branch less as much as possible 

Simple code is usually faster to execute

File: src/Common/serde.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+{
+    /// By default use `Revision` for version management, unless it has its own versions
+    if constexpr (HasGetVersionFromRevision<S>)
+        s.serialize(wb, static_cast<VersionType>(s.getVersionFromRevision(ProtonRevision::getVersionRevision())));
+    else
+        s.serialize(wb, static_cast<VersionType>(ProtonRevision::getVersionRevision()));
+}
+
+template <typename RB, VersionedDeserializable<RB> S>
+void ALWAYS_INLINE deserialize(S & s, RB && rb)
```
Comment (chenziliang) [inline src/Common/serde.h]: It seems we have a `&&` and we reading data from  it.

I still think we probably have too much complexity here and hence lots of slowness and confusion

File: src/Common/serde.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+template <typename T>
+concept HasGetVersionFromRevision = requires(T && a) {
+    {
+        a.getVersionFromRevision(ProtonRevision::getVersionRevision())
+    } -> std::convertible_to<VersionType>;
+};
+
+template <typename WB, VersionedSerializable<WB> S>
+void ALWAYS_INLINE serialize(const S & s, WB && wb)
```
Comment (chenziliang) [inline src/Common/serde.h]: it seems weird we have a && and we are writing data to it.

File: src/Common/serde.h (hunk 2 above)
Comment (yl-lisen) [reply src/Common/serde.h]: 1) change `&&` to `&`, it's make sense

2) I think it's no slowness, and no confusion, which just used to unify the serialization/deserialization interface to simplify the interface call of version management objects

Comment (yl-lisen) [reply src/Common/serde.h]: It's a little complicated, I'll readjust it

**Review by sunset3000 — APPROVED** [review]

**Pull request merged**